ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/chain/DefaultMutableBlockchain.java

```diff
@@ -192,18 +229,25 @@ private UInt256 calculateTotalDifficulty(final Block block) {
   private BlockAddedEvent updateCanonicalChainData(
       final BlockchainStorage.Updater updater,
       final Block newBlock,
-      final UInt256 totalDifficulty) {
+      final UInt256 totalDifficulty,
+      final boolean isFastSync) {
     final Hash chainHead = blockchainStorage.getChainHead().orElse(null);
     if (newBlock.getHeader().getNumber() != BlockHeader.GENESIS_BLOCK_NUMBER && chainHead == null) {
       throw new IllegalStateException("Blockchain is missing chain head.");
     }
 
     final Hash newBlockHash = newBlock.getHash();
     try {
-      if (chainHead == null || newBlock.getHeader().getParentHash().equals(chainHead)) {
+      if (chainHead == null
+          || newBlock.getHeader().getParentHash().equals(chainHead)
+          || isFastSync) {
         // This block advances the chain, update the chain head
         updater.putBlockHash(newBlock.getHeader().getNumber(), newBlockHash);
-        updater.setChainHead(newBlockHash);
+        if (isFastSync) {
```

Comment:
You should check that the new fast-sync block is connected to the previous fast-sync block or chainhead.

Reply:
We actually need to support re-orgs on the fast sync chain because we may need to pick a different pivot block (especially at the moment where restarting will definitely do that).  I imagine it's extremely rare to get a re-org that far back but better to handle it cleanly.

It deliberately doesn't track forks in fast sync - whatever chain we're currently importing is the one that's considered canonical because that's the one that will get us to the pivot block and have a world state matched up.